packages/react/src/ReactMemo.js

```diff
@@ -36,9 +36,14 @@ export function memo<Props>(
         return ownName;
       },
       set: function(name) {
-        ownName = name;
-        if (type.displayName == null) {
+        if (typeof name === 'string') {
+          ownName = name;
           type.displayName = name;
+        } else {
+          console.error(
+            "%s: is not valid displayName type, React memo's displayName should be a string",
+            typeof name,
+          );
```

Comment:
Why not follow the exact behavior of `forwardRef` here?
https://github.com/iChenLei/react/blob/d38a651fbea3bae55b572afbae1f96e836dd27f6/packages/react/src/ReactForwardRef.js#L51-L64

Specifically, only set `type.displayName = name` if it's not already `type.displayName === null`.

Reply:
https://github.com/facebook/react/blob/a0d6b155dc3a182162b9f91baac33df39d7919df/packages/react-devtools-shared/src/backend/renderer.js#L379-L392

If we set ownName we get the ownName(`ownName`) as displayName and otherwise we use type.render's displayName or name.

https://github.com/facebook/react/blob/a0d6b155dc3a182162b9f91baac33df39d7919df/packages/react-devtools-shared/src/backend/renderer.js#L414-L419

Yep, you are right, we need change the forwardRef logic as same as memoComponent.

Reply:
We just want allow user set displayName multi time, the common scenario is that user use third party react library such as mobx-react, mobx-react use hoc to wrap component and use `memo` to optimize rende performance. and now user can't set custom displayName to debug . for example (https://github.com/mobxjs/mobx/issues/2721

Reply:
My point was that I don't understand why we were making the implementation of `forwardRef` and `memo` different.

TBH I'm not sure that _either one_ modifying `type.displayName` is a good idea b'c that's a side effect. We do it because it helps with component stacks but I'm not sure that's the best workaround there either.

For instance:

```js
function MyComponent() {
  // ...
}

const MyMemoizedComponent = React.memo(MyComponent);
MyMemoizedComponent.displayName = 'Blah';

// Now if I render <MyComponent> it will also have a displayName of "blah"
// That doesn't seem good.
```

It's true that `forwardRef` and `memo` currently have different behaviors in component stacks and that's not good either. To me, this indicates that the right place to make this change is where we read the names rather than here. So I'm talking about `getComponentNameFromType` 😄 

Maybe it would be clearer for me to post a PR with an alternate idea for this.

Reply:
See #21392